### DISCUSSION TITLE
UCB and EI giving same trajectory in exploration

### DISCUSSION THREAD

**Discussion by utkarshp1161** [ziatdinovmax/gpax#64] [General]
Hi,

I am using gpax mostly now in hAE(https://github.com/utkarshp1161/hAE) --> main_hAE_gpax.py

I am facing one issue. The EI and UCB acquisition function are tracing the same trajectory in exploration. This didn't happen when I was using DKL from atomAI.

![post_forensics6_trajectory](https://github.com/ziatdinovmax/gpax/assets/52690126/7ebb24e6-38fc-4c86-a715-5834214c473d)

**Comment by ziatdinovmax** [comment]
Hmm... Can you please share a minimal reproducible code sample?

**Reply by utkarshp1161** [reply]
Thanks,

Here is the data:  !wget https://www.dropbox.com/s/nicyvttpihzh1cd/Plasmonic_sets_7222021_fixed.npy

The related minimal script( sorry still quite long):

import os
import matplotlib.gridspec as gridspec
from matplotlib.patches import Rectangle
import numpy as np
import matplotlib.pyplot as plt
import atomai as aoi
from atomai.utils import get_coord_grid, extract_patches_and_spectra

import logging
from datetime import datetime
current_time = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
filename = f"rf_gpax/log_{current_time}.log"

logging.basicConfig(filename=filename, level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s', datefmt='%Y-%m-%d %H:%M:%S')
logging.info("This is a test log message.")
import sys
import hAE
from hAE.scalarizer_eels import calculate_peaks, calculate_peaks_and_fit_gaussians_with_lmfit, calculate_peaks_and_fit_gaussians_with_lmfit_integral
from  hAE.all_dkl import dkl_explore
NPs = np.load("/nfs/home/upratius/scratch/projects/edited_AutomatedExperiment_Summer2023/utkarsh_eels/Plasmonic_sets_7222021_fixed.npy", allow_pickle=True).tolist()

key = 3

save_dir = "/nfs/home/upratius/scratch/projects/edited_AutomatedExperiment_Summer2023/utkarsh_eels/hAE/"
specim   = NPs[f'{key}']['spectrum image']
eax      = NPs[f'{key}']['energy axis']
pxsizenm = NPs[f'{key}']['scale']
img      = NPs[f'{key}']['image']

plt.imshow(img)
plt.savefig(save_dir + "1_img.png")

window_size       = 8          
spectralavg       = 4            

coordinates = get_coord_grid(img, step = 1, return_dict=False)
patches, spectra, coords = extract_patches_and_spectra(specim, img, coordinates=coordinates, window_size=window_size, avg_pool=spectralavg)

ws = window_size 

full_img    = np.copy(img)
indices     = np.copy(coords)
features    = np.copy(patches)
targets     = np.copy(spectra)
scale       = np.copy(pxsizenm)

e_axis      = np.copy(eax[::spectralavg])

while e_axis.shape[0] > spectra[0].shape[0]:
  e_axis = e_axis[:-1]

k = 350 

band = [0.3, 0.4]
peaks_all_scalar1, features_all, indices_all = calculate_peaks(targets, features, indices, e_axis, band)

y = np.copy(peaks_all_scalar1)
n, d1, d2 = features.shape
X = features.reshape(n, d1*d2)

logging.info("Reached active exp----------------------------------")

exploration_steps = 100
num_cycles        = 100

xi = 0.01  
cm = 'viridis'
shrink = 0.7

ts = 0.995  
rs = 42   

acq = 3  

rf = save_dir + "rf_gpax/"  
if not os.path.exists(rf):
    os.mkdir(rf)

acq_funcs = ['EI', 'MU', 'UCB'] 

if acq == 3:             
  if not os.path.exists(rf + acq_funcs[0]):
    os.mkdir(rf + acq_funcs[0])
  if not os.path.exists(rf + acq_funcs[1]):
    os.mkdir(rf + acq_funcs[1])
  if not os.path.exists(rf + acq_funcs[2]):
    os.mkdir(rf + acq_funcs[2])
elif acq < 3:
  if not os.path.exists(rf + acq_funcs[acq]):  
    os.mkdir(rf + acq_funcs[acq])

logging.info("Reached dkl explore----------------------------------")

save_explore =  "/explore_dkl_record/"
logging.info("Reached dkl explore band 1----------------------------------")

y_targ = peaks_all_scalar1    
y_targ = (y_targ-y_targ.min())/(y_targ.max()-y_targ.min()) 
y = y_targ.reshape(-1)

if acq < 3:   
  acq_idx = acq
  dkl_explore(X, y, indices_all, ts, rs, rf, acq_funcs, exploration_steps, acq, acq_idx, ws, img, window_size, xi, save_explore, num_cycles)

elif acq == 3:  
  for i in range (3):
    acq_idx = i
    dkl_explore(X, y, indices_all, ts, rs, rf, acq_funcs, exploration_steps,acq, acq_idx, ws, img,

**Comment by ziatdinovmax** [comment]
@utkarshp1161 - does running a **pure GPax** implementation of DKL on this dataset with UCB and EI also result in identical answers?

**Comment by utkarshp1161** [comment]
Hi Maxim,

I have tried the UCB and EI in https://github.com/utkarshp1161/hAE/blob/main/notebooks/gpax/debug_acq_fn.ipynb

Still both are giving the same exploration.

Will it be possible for you to try it out?

Here is the data: !wget https://www.dropbox.com/s/nicyvttpihzh1cd/Plasmonic_sets_7222021_fixed.npy
Please install : pip install git+https://github.com/utkarshp1161/hAE in the python env

**Comment by ziatdinovmax** [comment]
@utkarshp1161 did you have a chance to check a *pure* GPax implementation (without hAE wrappers) of the same acquisition functions? We need to be able to distinguish whether the bug causing this lives in GPax itself or hAE.
